File: 2023/15/part1.ts
import { solveCalendar } from '../util/solveCalendar'
import { writeOutput } from '../util/writeOutput'

type Step = string[]

function parseSteps(input: string[]): Step[] {
    const line = input.join('')
    const steps: Step[] = line.split(',').map((str) => str.split(''))
    return steps
}

function ASCII(char: string) {
    return char.charCodeAt(0)
}

function HASH(step: Step) {
    let currentValue = 0
    for (const char of step) {
        currentValue += ASCII(char)
        currentValue *= 17
        currentValue %= 256
    }
    return currentValue
}

function solver(input: string[]): number {
    const output: string[] = []

    const steps = parseSteps(input)

    let sum = 0
    for (const step of steps) {
        const hash = HASH(step)
        sum += hash
        output.push(`${step.join('').padEnd(8)} - ${hash.toString().padStart(3)}`)
    }

    writeOutput({ day: 15, part: 1, output })
    return sum
}

solveCalendar({
    day: 15,
    sampleAnswer: 1320,
    solver, // 517965 ⭐️
})

File: 2024/src/03/part1.py
from ..util.problem_solver import problem_solver
import re
from typing import Tuple


def solver(lines: list[str]) -> int:
    input = "".join(lines)
    sum = 0
    for a, b in find_multiplication_pairs(input):
        sum += a * b
    return sum


def find_multiplication_pairs(s: str) -> list[Tuple[int, int]]:
    str_nums: list[Tuple[str, str]] = re.findall(r"mul\(([0-9]{1,3}),([0-9]{1,3})\)", s)
    return [(int(a), int(b)) for a, b in str_nums]


if __name__ == "__main__":
    problem_solver(
        day=3,
        solver=solver,
        sample_nr=1,
        sample_answer=161,
    )
